File: src/js/graphs/test_bfs.mjs
import { readFileSync } from 'fs';
import { bfs } from './bfs.mjs';

const size = process.argv[2] || 'small';
const path = `../../../datasets/graphs/${size}.bin`;

const buffer = readFileSync(path);
const numOfNodes = buffer.readInt32LE(0);
const numOfEdges = buffer.readInt32LE(4);

const from = new Int32Array(buffer.buffer, buffer.byteOffset + 8, numOfEdges);
const to = new Int32Array(buffer.buffer, buffer.byteOffset + 8 + (numOfEdges * 4), numOfEdges);

// Build CSR structure
const offsets = new Int32Array(numOfNodes + 1);
const neighbors = new Int32Array(numOfEdges);

// Count edges for each node to determine offset sizes
const counts = new Int32Array(numOfNodes);
for (let i = 0; i < numOfEdges; i++) {
    counts[from[i]]++;
}

// Prefix sum to calculate offsets
offsets[0] = 0;
for (let i = 0; i < numOfNodes; i++) {
    offsets[i + 1] = offsets[i] + counts[i];
}

// Fill the neighbors array using the offsets
const cursor = new Int32Array(numOfNodes);
for (let i = 0; i < numOfEdges; i++) {
    const startNode = from[i];
    const destinationNode = to[i];
    const writePos = offsets[startNode] + cursor[startNode];
    neighbors[writePos] = destinationNode;
    cursor[startNode]++;
}

const graphData = { numOfNodes, offsets, neighbors };
const visited = new Int32Array(numOfNodes);
const dist = new Int32Array(numOfNodes);

bfs(graphData, 0, visited, dist);

const reachable = visited.reduce((acc, v) => acc + v, 0);
const maxDist = dist.reduce((acc, d) => d > acc ? d : acc, 0);

console.log('------------');
console.log('BFS');
console.log('------------');
console.log(`Dataset: ${path}`);
console.log(`Loaded: ${numOfNodes} nodes, ${numOfEdges} edges`);
console.log(`BFS from node 0:`);
console.log(`Reachable nodes : ${reachable} / ${numOfNodes}`);
console.log(`Max distance    : ${maxDist}`);
console.log(`dist[0]         : ${dist[0]}`);
console.log(`dist[1]         : ${dist[1]}`);
console.log(`dist[2]         : ${dist[2]}`);
console.log(`dist[3]         : ${dist[3]}`);

File: src/c/graphs/test_bfs.c
#include <stdio.h>
#include <stdlib.h>
#include "bfs.h"
#include "../utils/utils.h"

static Graph *build_graph(const GraphData *gd) {
    Graph *g = graph_create(gd->num_nodes);
    graph_build(g, gd->num_edges, gd->from, gd->to);
    return g;
}

int main(int argc, char *argv[]) {
    const char *pathType = "small";
    if (argc >= 2) pathType = argv[1];

    char path[256];
    snprintf(path, sizeof(path), "../../../datasets/graphs/%s.bin", pathType);

    GraphData *gd = load_graph_data(path);
    printf("Dataset: %s\n", path);
    printf("Loaded: %d nodes, %d edges\n", gd->num_nodes, gd->num_edges);

    Graph *g = build_graph(gd);

    int *visited = (int *)malloc(g->num_nodes * sizeof(int));
    int *dist    = (int *)malloc(g->num_nodes * sizeof(int));

    bfs(g, 0, visited, dist);

    int reachable = 0, max_dist = 0;
    for (int i = 0; i < g->num_nodes; i++) {
        if (visited[i]) {
            reachable++;
            if (dist[i] > max_dist) max_dist = dist[i];
        }
    }

    printf("BFS from node 0:\n");
    printf("  Reachable nodes : %d / %d\n", reachable, g->num_nodes);
    printf("  Max distance    : %d\n", max_dist);
    printf("  dist[0]         : %d\n", dist[0]);
    printf("  dist[1]         : %d\n", dist[1]);
    printf("  dist[2]         : %d\n", dist[2]);
    printf("  dist[3]         : %d\n", dist[3]);

    free(visited);
    free(dist);
    graph_free(g);
    free_graph(gd);
    return 0;
}
